jeos-firstboot-enroll

```diff
@@ -12,197 +18,316 @@ FDE_SEAL_PCR_LIST="0,2,4,7,9"
 
 have_luks2()
 {
-    [ "${#luks2_devices[@]}" -gt 0 ]
+	[ "${#luks2_devices[@]}" -gt 0 ]
 }
 
 detect_luks2()
 {
-    local dev fstype
-    [ -z "$luks2_devices" ] || return 0
-    while read -r dev fstype; do
-	[ "$fstype" = 'crypto_LUKS' ] || continue
-	cryptsetup isLuks --type luks2 "$dev" || continue
-	luks2_devices+=("$dev")
-    done < <(lsblk --noheadings -o PATH,FSTYPE)
-    have_luks2
+	local dev fstype
+	[ -z "$luks2_devices" ] || return 0
+	while read -r dev fstype; do
+		[ "$fstype" = 'crypto_LUKS' ] || continue
+		cryptsetup isLuks --type luks2 "$dev" || continue
+		luks2_devices+=("$dev")
+	done < <(lsblk --noheadings -o PATH,FSTYPE)
+	have_luks2
 }
 
 enroll_systemd_firstboot() {
-    [ -e /usr/bin/systemd-cryptenroll ] || return 0
-    detect_luks2 || return 0
-
-    local has_fido2=${JEOS_HAS_FIDO2:-}
-    local has_tpm2=
-
-    [ -z "$(systemd-cryptenroll --fido2-device=list 2>/dev/null)" ] || has_fido2=1
-    if [ -e '/sys/class/tpm/tpm0' ]; then
-	if have_pcrlock; then
-	    has_tpm2=lock
-	elif have_pcr_oracle; then
-	    has_tpm2=oracle
+	[ -e /usr/bin/systemd-cryptenroll ] || return 0
+	crypt_keyid="$(keyctl id %user:cryptenroll)"
+	[ -n "$crypt_keyid" ] || return 0
+
+	detect_luks2 || return 0
+
+	welcome_screen_with_console_switch
+
+	local has_fido2=${JEOS_HAS_FIDO2:-}
+	local has_tpm2=
+
+	[ -z "$(systemd-cryptenroll --fido2-device=list 2>/dev/null)" ] || has_fido2=1
+	if [ -e '/sys/class/tpm/tpm0' ]; then
+		if have_pcrlock && ! is_pcr_oracle; then
+			has_tpm2=lock
+		elif have_pcr_oracle; then
+			has_tpm2=oracle
+		fi
 	fi
-    fi
-
-   # For now seems that if a FIDO2 key is enrolled, it will take
-   # precedence over the TPM2 and the key will be asked to be present
-   # in subsequent boots.
-   if [ "$has_fido2" = '1' ] && [ -n "$has_tpm2" ]; then
-	local list=('FIDO2' 'FIDO2' 'TPM2' 'TPM2' 'none' $"Skip")
-	d --no-tags --default-item 'FIDO2' --menu $"Select unlock device" 0 0 "$(menuheight ${#list[@]})" "${list[@]}"
-	[ "$result" = 'FIDO2' ] && with_fido2=1
-	[ "$result" = 'TPM2' ] && with_tpm2="$has_tpm2"
-    elif [ "$has_fido2" ]; then
-	dialog $dialog_alternate_screen --backtitle "$PRETTY_NAME" --yesno $"Unlock encrypted disk via FIDO2 token?" 0 0 && with_fido2=1
-    elif [ -n "$has_tpm2" ]; then
-	dialog $dialog_alternate_screen --backtitle "$PRETTY_NAME" --yesno $"Unlock encrypted disk via TPM?" 0 0 && with_tpm2="$has_tpm2"
-    fi
-    return 0
+
+	while true; do
```

Comment:
Here start the interactive part, that now is not trivial.

Ideally all this should be unattended (the typical MicroOS use case will be in servers, with multiple installations). For a PoC was OK always to require interaction, but we should move out from it now.

In the same way that rd.credentials was added for this set, can we add something like rd.interactive to enter this codepath? If is missing and the TPM2 is detected, we can enroll it automatically.

Reply:
Huh? This is the firstboot *wizard*. The whole point of this thing is to be interactive. The autodeployment case is meant to be handled by combustion. When combustion is used, no firstboot wizard is started.

Having said that I've also implemented credential support for the wizard which can actually make it non-interactive: https://github.com/openSUSE/jeos-firstboot/pull/119
Once that's in we can implement credential support also for the enroll module.

Reply:
> Huh? This is the firstboot wizard. 

wizard is a too strong word for the jeos firstboot modules ...

> The autodeployment case is meant to be handled by combustion. When combustion is used, no firstboot wizard is started.

OK then.